dvc/repo/run.py

```diff
@@ -5,18 +5,26 @@
 @locked
 @scm_context
 def run(self, no_exec=False, **kwargs):
-    from dvc.stage import Stage
+    from dvc.stage import PipelineStage, Stage
 
-    stage = Stage.create(self, **kwargs)
+    stage_cls = PipelineStage if kwargs.get("name") else Stage
+    stage = stage_cls.create(self, **kwargs)
+    if stage_cls == PipelineStage:
+        stage.name = kwargs["name"]
 
     if stage is None:
         return None
 
-    self.check_modified_graph([stage])
+    # TODO: check if the stage with given name already exists, don't allow that
+    self.check_modified_graph([stage], self.pipeline_stages)
+    self.pipeline_stages.append(stage)
 
     if not no_exec:
         stage.run(no_commit=kwargs.get("no_commit", False))
 
+    if stage_cls == PipelineStage:
+        stage.dvcfile.dump_multistages(stage, stage.path)
```

Comment:
This is really strange. Why not
```
dvcfile.dump()
```
? Just let the dvcfile know what type it is, no reason for us to do such hacks.

Reply:
Yup, but we don't really need to dump to dvcfile at all (except `dvc run` obviously). This PR needs a lot of adjustments after #3619.